### PR TITLE
Ensure that all variants have a description

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR includes several changes to the `variants` directive:

- [x] Add audit test for variant descriptions
- [x] Add type hints
- [x] Fix examples of `default="True"` and `default="False"`

Not 100% confident that these type hints are correct. Also don't know what the return type should be.

### CODE REVIEW

**Comment by jacobrking** [comment]
> @hguo @jacobrking @jmrosinski @streichler can you review this PR?
> 
> This PR modifies the following package(s), for which you are listed as a maintainer:
> 
> * armcomputelibrary
> * autodiff
> * davix
> * freesasa
> * ftk
> * g2c
> * generax
> * gptl
> * gsl-lite
> * hipblas
> * hipcub
> * hipfft
> * hiprand
> * hipsolver
> * hipsparse
> * kokkos
> * kokkos-kernels
> * legion
> * libxcrypt
> * mgard
> * miopen-tensile
> * nimrod-aai
> * nlcglib
> * octave
> * py-hist
> * py-openpmd-viewer
> * rccl
> * rocalution
> * rocblas
> * rocfft
> * rocm-tensile
> * rocprim
> * rocrand
> * rocsolver
> * rocsparse
> * rocthrust
> * rocwmma
> * root
> * semiprof
> * spglib
> * tandem
> * tiled-mm
> * tmalign
> * tmscore
> * trilinos
> * ucx
> * unifyfs
> * upp
> * usalign
> * wps
> * wrf
> * xyce

LGTM

**Review by psakievich — APPROVED** [review]
Approve the trilinos change

**Review by climbfuji — APPROVED** [review]
Approving gsl-lite changes, thanks for the PR!

**Review by alalazo — APPROVED** [review]

**Review by alalazo — CHANGES_REQUESTED** [review]
Current changes LGTM. We need to change the mock repositories too though.

**Comment by hguo** [comment]
I approve the changes for FTK.

**Review by HadrienG2 — APPROVED** [review]

File: var/spack/repos/builtin/packages/mgard/package.py [hunk 1]
```diff
@@ -26,10 +26,15 @@ class Mgard(CMakePackage, CudaPackage):
     version("2021-11-12", commit="3c05c80a45a51bb6cc5fb5fffe7b1b16787d3366")
     version("2020-10-01", commit="b67a0ac963587f190e106cc3c0b30773a9455f7a")
 
-    variant("serial", when="@2022-11-18:", default=True)
-    variant("openmp", when="@2022-11-18:", default=True)
-    variant("timing", when="@2022-11-18:", default=False)
-    variant("unstructured", when="@2022-11-18:", default=False)
+    variant("serial", when="@2022-11-18:", default=True, description="Enable SERIAL support")
```
Comment (robertu94) [inline var/spack/repos/builtin/packages/mgard/package.py]: enable the classic non-parallel implmementation

File: var/spack/repos/builtin/packages/mgard/package.py [hunk 2]
```diff
@@ -26,10 +26,15 @@ class Mgard(CMakePackage, CudaPackage):
...
+    variant("openmp", when="@2022-11-18:", default=True, description="Enable OpenMP support")
+    variant("timing", when="@2022-11-18:", default=False, description="Enable timing")
```
Comment (robertu94) [inline var/spack/repos/builtin/packages/mgard/package.py]: enable profile timings

**Review by robertu94 — CHANGES_REQUESTED** [review]
Two minor tweaks.  I don't know why, but the web interface didn't allow me to "suggest" the changes this time.

**Comment by adamjstewart** [comment]
I just copied the descriptions from CMake but yours sound better.

**Review by MichaelLaufer — APPROVED** [review]

File: lib/spack/spack/directives.py [hunk 3]
```diff
@@ -663,39 +663,36 @@ def _execute_patch(pkg_or_dep):
...
-    description="",
-    values=None,
-    multi=None,
-    validator=None,
-    when=None,
-    sticky=False,
+    name: str,
+    *,
+    default: Optional[Any] = None,
+    description: str,
```
Comment (haampie) [inline lib/spack/spack/directives.py]: I get the idea, but would recommend not to make unnecessary breaking changes. If you really care about descriptions, we can add it to audits. But like this it's just more painful to upgrade spack with custom repos.

The fact that you now need to add `description=""` in tons of places tells a lot ;)

**Review by haampie — CHANGES_REQUESTED** [review]
Not a fan of unnecessary breaking API changes

File: lib/spack/spack/directives.py [hunk 4]
```diff
@@ -663,39 +663,36 @@ def _execute_patch(pkg_or_dep):
...
+    values: Optional[Union[Tuple[Any, ...], Callable[[Any], bool]]] = None,
```
Comment (haampie) [inline lib/spack/spack/directives.py]: I think you want to use `collections.abc.Sequence` here to support `values=any_combination_of(...)`

File: lib/spack/spack/directives.py (hunk 3 above)
Comment (adamjstewart) [reply lib/spack/spack/directives.py]: I'm not opposed to changing it to an audit. Let me dig around and see how hard that would be.

**Review by haampie — APPROVED** [review]

File: lib/spack/spack/directives.py [hunk 5]
```diff
@@ -663,39 +663,35 @@ def _execute_patch(pkg_or_dep):
...
-    description="",
-    values=None,
-    multi=None,
-    validator=None,
-    when=None,
-    sticky=False,
+    name: str,
+    default: Optional[Any] = None,
+    description: str = "",
+    values: Optional[Union[collections.abc.Sequence, Callable[[Any], bool]]] = None,
```
Comment (alalazo) [inline lib/spack/spack/directives.py:669]: This one misses the `DisjointSet...` etc. but we can add it later

**Review by alalazo — APPROVED** [review]

**Review by robertu94 — APPROVED** [review]

**Pull request merged**